### ISSUE TITLE
Configuration Error

### ISSUE THREAD

**Issue by aardelean** [GoogleCloudPlatform/fluent-plugin-detect-exceptions#39]
Hi, I am trying to use the detect-exceptions in a docker environment configured with fluentd driver, but my configuration throws some errors in the fluentd output:
```
2018-06-01 15:33:22 +0000 [info]: parsing config file is succeeded path="/fluentd/etc/conf/docker.conf"
2018-06-01 15:33:22 +0000 [info]: parsing config file is succeeded path="/fluentd/etc/conf/docker.conf"
/usr/lib/ruby/2.3.0/rubygems/core_ext/kernel_require.rb:127:in 'require': cannot load such file -- fluent/plugin/exception_detector (LoadError)
	from /usr/lib/ruby/2.3.0/rubygems/core_ext/kernel_require.rb:127:in 'rescue in require'
	from /usr/lib/ruby/2.3.0/rubygems/core_ext/kernel_require.rb:40:in 'require'
	from /var/lib/gems/2.3.0/gems/fluent-plugin-detect-exceptions-0.0.11/lib/fluent/plugin/out_detect_exceptions.rb:16:in '<top (required)>'
	from /usr/lib/ruby/2.3.0/rubygems/core_ext/kernel_require.rb:55:in 'require'
	from /usr/lib/ruby/2.3.0/rubygems/core_ext/kernel_require.rb:55:in 'require'
	from /var/lib/gems/2.3.0/gems/fluentd-1.2.1/lib/fluent/registry.rb:102:in 'block in search'
	from /var/lib/gems/2.3.0/gems/fluentd-1.2.1/lib/fluent/registry.rb:99:in 'each'
	from /var/lib/gems/2.3.0/gems/fluentd-1.2.1/lib/fluent/registry.rb:99:in 'search'
	from /var/lib/gems/2.3.0/gems/fluentd-1.2.1/lib/fluent/registry.rb:44:in 'lookup'
	from /var/lib/gems/2.3.0/gems/fluentd-1.2.1/lib/fluent/plugin.rb:146:in 'new_impl'
	from /var/lib/gems/2.3.0/gems/fluentd-1.2.1/lib/fluent/plugin.rb:104:in 'new_output'
	from /var/lib/gems/2.3.0/gems/fluentd-1.2.1/lib/fluent/agent.rb:128:in 'add_match'
	from /var/lib/gems/2.3.0/gems/fluentd-1.2.1/lib/fluent/agent.rb:72:in 'block in configure'
	from /var/lib/gems/2.3.0/gems/fluentd-1.2.1/lib/fluent/agent.rb:64:in 'each'
	from /var/lib/gems/2.3.0/gems/fluentd-1.2.1/lib/fluent/agent.rb:64:in 'configure'
	from /var/lib/gems/2.3.0/gems/fluentd-1.2.1/lib/fluent/root_agent.rb:112:in 'configure'
	from /var/lib/gems/2.3.0/gems/fluentd-1.2.1/lib/fluent/engine.rb:131:in 'configure'
	from /var/lib/gems/2.3.0/gems/fluentd-1.2.1/lib/fluent/engine.rb:96:in 'run_configure'
	from /var/lib/gems/2.3.0/gems/fluentd-1.2.1/lib/fluent/supervisor.rb:795:in 'run_configure'
	from /var/lib/gems/2.3.0/gems/fluentd-1.2.1/lib/fluent/supervisor.rb:579:in 'dry_run'
	from /var/lib/gems/2.3.0/gems/fluentd-1.2.1/lib/fluent/supervisor.rb:597:in 'supervise'
	from /var/lib/gems/2.3.0/gems/fluentd-1.2.1/lib/fluent/supervisor.rb:502:in 'run_supervisor'
	from /var/lib/gems/2.3.0/gems/fluentd-1.2.1/lib/fluent/command/fluentd.rb:310:in '<top (required)>'
	from /usr/lib/ruby/2.3.0/rubygems/core_ext/kernel_require.rb:55:in 'require'
	from /usr/lib/ruby/2.3.0/rubygems/core_ext/kernel_require.rb:55:in 'require'
	from /var/lib/gems/2.3.0/gems/fluentd-1.2.1/bin/fluentd:8:in'<top (required)>'
	from /usr/local/bin/fluentd:22:in 'load'
	from /usr/local/bin/fluentd:22:in '<main>'
```

and the configuration is :
```
<source>
  @type forward
  port 24224
  tag removeme
  bind 0.0.0.0
</source>

<filter **>
  @type record_transformer
  enable_ruby
  <record>
    timestamp ${ require 'time'; Time.now.utc.iso8601(3) }
  </record>
  remove_keys source container_id
</filter>

<match removeme.*>
  @type detect_exceptions
  message log
  languages java
  remove_tag_prefix removeme
  multiline_flush_interval 0.5
</match>

<match *>
  @type elasticsearch
  host elasticsearch
  port 9200
  time_key timestamp
  index_name fluentb
  type_name fluentb
</match>
```

**Comment by igorpeshansky** [comment]
I strongly suspect it's because of https://github.com/GoogleCloudPlatform/fluent-plugin-detect-exceptions/issues/32#issuecomment-389621089. We will investigate.

**Comment by aardelean** [comment]
@igorpeshansky thanks for your fast response. How can I somehow fix it or test if this is the real issue?

**Comment by igorpeshansky** [comment]
You can find `exception_detector.rb` under `/opt/google-fluentd/embedded` and `chmod a+r`. If that fixes it, then it's as I suspected.

**Comment by aardelean** [comment]
I have installed the plugin with sudo gem install fluentd-plugin-detect-exceptions in the Dockerfile. Is there another way?

**Comment by aardelean** [comment]
@igorpeshansky I have tried to grant file permissions with chmod a+x -R fluent/ and still get the same error. I doubt it is the permissions. Is there another version I could try?

**Comment by gotttlieb** [comment]
I think that i am getting the same error here:

```2018-06-14 11:28:53 +0000 [info]: parsing config file is succeeded path="/fluentd/etc/fluent.conf"
2018-06-14 11:28:53 +0000 [info]: 'flush_interval' is configured at out side of <buffer>. 'flush_mode' is set to 'interval' to keep existing behaviour
/usr/lib/ruby/2.3.0/rubygems/core_ext/kernel_require.rb:127:in `require': cannot load such file -- fluent/plugin/exception_detector (LoadError)
        from /usr/lib/ruby/2.3.0/rubygems/core_ext/kernel_require.rb:127:in `rescue in require'
        from /usr/lib/ruby/2.3.0/rubygems/core_ext/kernel_require.rb:40:in `require'
        from /var/lib/gems/2.3.0/gems/fluent-plugin-detect-exceptions-0.0.11/lib/fluent/plugin/out_detect_exceptions.rb:16:in `<top (required)>'
        from /usr/lib/ruby/2.3.0/rubygems/core_ext/kernel_require.rb:55:in `require'
        from /usr/lib/ruby/2.3.0/rubygems/core_ext/kernel_require.rb:55:in `require'
        from /var/lib/gems/2.3.0/gems/fluentd-1.2.1/lib/fluent/registry.rb:102:in `block in search'
        from /var/lib/gems/2.3.0/gems/fluentd-1.2.1/lib/fluent/registry.rb:99:in `each'
        from /var/lib/gems/2.3.0/gems/fluentd-1.2.1/lib/fluent/registry.rb:99:in `search'
        from /var/lib/gems/2.3.0/gems/fluentd-1.2.1/lib/fluent/registry.rb:44:in `lookup'
        from /var/lib/gems/2.3.0/gems/fluentd-1.2.1/lib/fluent/plugin.rb:146:in `new_impl'
        from /var/lib/gems/2.3.0/gems/fluentd-1.2.1/lib/fluent/plugin.rb:104:in `new_output'
        from /var/lib/gems/2.3.0/gems/fluentd-1.2.1/lib/fluent/agent.rb:128:in `add_match'
        from /var/lib/gems/2.3.0/gems/fluentd-1.2.1/lib/fluent/agent.rb:72:in `block in configure'
        from /var/lib/gems/2.3.0/gems/fluentd-1.2.1/lib/fluent/agent.rb:64:in `each'
        from /var/lib/gems/2.3.0/gems/fluentd-1.2.1/lib/fluent/agent.rb:64:in `configure'
        from /var/lib/gems/2.3.0/gems/fluentd-1.2.1/lib/fluent/root_agent.rb:112:in `configure'
        from /var/lib/gems/2.3.0/gems/fluentd-1.2.1/lib/fluent/engine.rb:131:in `configure'
        from /var/lib/gems/2.3.0/gems/fluentd-1.2.1/lib/fluent/engine.rb:96:in `run_configure'
        from /var/lib/gems/2.3.0/gems/fluentd-1.2.1/lib/fluent/supervisor.rb:795:in `run_configure'
        from /var/lib/gems/2.3.0/gems/fluentd-1.2.1/lib/fluent/supervisor.rb:579:in `dry_run'
        from /var/lib/gems/2.3.0/gems/fluentd-1.2.1/lib/fluent/supervisor.rb:597:in `supervise'
        from /var/lib/gems/2.3.0/gems/fluentd-1.2.1/lib/fluent/supervisor.rb:502:in `run_supervisor'
        from /var/lib/gems/2.3.0/gems/fluentd-1.2.1/lib/fluent/command/fluentd.rb:310:in `<top (required)>'
        from /usr/lib/ruby/2.3.0/rubygems/core_ext/kernel_require.rb:55:in `require'
        from /usr/lib/ruby/2.3.0/rubygems/core_ext/kernel_require.rb:55:in `require'
        from /var/lib/gems/2.3.0/gems/fluentd-1.2.1/bin/fluentd:8:in `<top (required)>'
        from /usr/local/bin/fluentd:22:in `load'
        from /usr/local/bin/fluentd:22:in `<main>'```

**Comment by aa23** [comment]
I'm having a similar issue as well on Centos7:

```
Jun 14 15:42:13 centos-7 systemd[1]: Starting td-agent: Fluentd based data collector for Treasure Data...
Jun 14 15:42:14 centos-7 fluentd[21157]: /opt/td-agent/embedded/lib/ruby/site_ruby/2.4.0/rubygems/core_ext/kernel_require.rb:133:in `require': cannot load such file -- fluent/plugin/exception_detector (LoadError)
Jun 14 15:42:14 centos-7 fluentd[21157]: from /opt/td-agent/embedded/lib/ruby/site_ruby/2.4.0/rubygems/core_ext/kernel_require.rb:133:in `rescue in require'
Jun 14 15:42:14 centos-7 fluentd[21157]: from /opt/td-agent/embedded/lib/ruby/site_ruby/2.4.0/rubygems/core_ext/kernel_require.rb:40:in `require'
Jun 14 15:42:14 centos-7 fluentd[21157]: from /opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/fluent-plugin-detect-exceptions-0.0.11/lib/fluent/plugin/out_detect_exceptions.rb:16:in `<top (required)>'
Jun 14 15:42:14 centos-7 fluentd[21157]: from /opt/td-agent/embedded/lib/ruby/site_ruby/2.4.0/rubygems/core_ext/kernel_require.rb:55:in `require'
Jun 14 15:42:14 centos-7 fluentd[21157]: from /opt/td-agent/embedded/lib/ruby/site_ruby/2.4.0/rubygems/core_ext/kernel_require.rb:55:in `require'
Jun 14 15:42:14 centos-7 fluentd[21157]: from /opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/fluentd-1.0.2/lib/fluent/registry.rb:102:in `block in search'
Jun 14 15:42:14 centos-7 fluentd[21157]: from /opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/fluentd-1.0.2/lib/fluent/registry.rb:99:in `each'
Jun 14 15:42:14 centos-7 fluentd[21157]: from /opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/fluentd-1.0.2/lib/fluent/registry.rb:99:in `search'
Jun 14 15:42:14 centos-7 fluentd[21157]: from /opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/fluentd-1.0.2/lib/fluent/registry.rb:44:in `lookup'
Jun 14 15:42:14 centos-7 fluentd[21157]: from /opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/fluentd-1.0.2/lib/fluent/plugin.rb:146:in `new_impl'
Jun 14 15:42:14 centos-7 fluentd[21157]: from /opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/fluentd-1.0.2/lib/fluent/plugin.rb:104:in `new_output'
Jun 14 15:42:14 centos-7 fluentd[21157]: from /opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/fluentd-1.0.2/lib/fluent/agent.rb:128:in `add_match'
Jun 14 15:42:14 centos-7 fluentd[21157]: from /opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/fluentd-1.0.2/lib/fluent/agent.rb:72:in `block in configure'
Jun 14 15:42:14 centos-7 fluentd[21157]: from /opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/fluentd-1.0.2/lib/fluent/agent.rb:64:in `each'
Jun 14 15:42:14 centos-7 fluentd[21157]: from /opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/fluentd-1.0.2/lib/fluent/agent.rb:64:in `configure'
Jun 14 15:42:14 centos-7 fluentd[21157]: from /opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/fluentd-1.0.2/lib/fluent/root_agent.rb:112:in `configure'
Jun 14 15:42:14 centos-7 fluentd[21157]: from /opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/fluentd-1.0.2/lib/fluent/engine.rb:131:in `configure'
Jun 14 15:42:14 centos-7 fluentd[21157]: from /opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/fluentd-1.0.2/lib/fluent/engine.rb:96:in `run_configure'
Jun 14 15:42:14 centos-7 fluentd[21157]: from /opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/fluentd-1.0.2/lib/fluent/supervisor.rb:770:in `run_configure'
Jun 14 15:42:14 centos-7 fluentd[21157]: from /opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/fluentd-1.0.2/lib/fluent/supervisor.rb:553:in `dry_run'
Jun 14 15:42:14 centos-7 fluentd[21157]: from /opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/fluentd-1.0.2/lib/fluent/supervisor.rb:571:in `supervise'
Jun 14 15:42:14 centos-7 fluentd[21157]: from /opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/fluentd-1.0.2/lib/fluent/supervisor.rb:476:in `run_supervisor'
Jun 14 15:42:14 centos-7 fluentd[21157]: from /opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/fluentd-1.0.2/lib/fluent/command/fluentd.rb:310:in `<top (required)>'
Jun 14 15:42:14 centos-7 fluentd[21157]: from /opt/td-agent/embedded/lib/ruby/site_ruby/2.4.0/rubygems/core_ext/kernel_require.rb:55:in `require'
Jun 14 15:42:14 centos-7 fluentd[21157]: from /opt/td-agent/embedded/lib/ruby/site_ruby/2.4.0/rubygems/core_ext/kernel_require.rb:55:in `require'
Jun 14 15:42:14 centos-7 fluentd[21157]: from /opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/fluentd-1.0.2/bin/fluentd:8:in `<top (required)>'
```

Config block:

```
<match docker.**>
  @type detect_exceptions  # https://github.com/GoogleCloudPlatform/fluent-plugin-detect-exceptions
  remove_tag_prefix foo
  message log
  languages all
  multiline_flush_interval 0.1
</match>
```

Commands used to install:

```
yum install -y make gcc ruby-devel rubygems-integration
curl -L https://toolbelt.treasuredata.com/sh/install-redhat-td-agent3.sh | sh
td-agent-gem install fluent-plugin-parser
```

**Comment by aa23** [comment]
As others have noted in this issue and others, I've resolved the error after plugin installation with the following commands:

```
[root@work-centos-7 ~]# ls -l /opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/fluent-plugin-detect-exceptions-0.0.11/lib/fluent/plugin/exception_detector.rb 
-rw-r-----. 1 root root 11514 Jun 14 15:41 /opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/fluent-plugin-detect-exceptions-0.0.11/lib/fluent/plugin/exception_detector.rb

[root@work-centos-7 ~]# chmod 644 /opt/td-agent/embedded/lib/ruby/gems/2.4.0/gems/fluent-plugin-detect-exceptions-0.0.11/lib/fluent/plugin/exception_detector.rb 
```

**Comment by grenzr** [comment]
Just hit the same problem whilst using the 'common' tag of a Fluentd image built from https://github.com/theasp/docker-fluentd-plugins

**Comment by igorpeshansky** [comment]
This should be fixed in [v0.0.12](https://github.com/GoogleCloudPlatform/fluent-plugin-detect-exceptions/releases/tag/v0.0.12). If not, please reopen.
